### PR TITLE
feat: spras_revision

### PR DESCRIPTION
**Pull request by tristan-f-r**
This adds the unique spras_revision to every single paramater combination (before hashing) and the dataset label, to provide OSDF support on the level of deterministic algorithms.

For non-deterministic algorithms, I would like #292.

### CODE REVIEW

**Comment by tristan-f-r** [comment]
I will need to restructure the way that the `summary.py` tests are done to test them automatically. In better news, that will close #165!

**Comment by tristan-f-r** [comment]
I'm having trouble getting the cytoscape test to pass: This works perfectly fine locally, but I have no clue where the cytoscape fail is coming from.

**Comment by tristan-f-r** [comment]
The move from `config.yaml` to `example.yaml` was nearly arbitrary (it's been renamed since it does actually differ from the top-level `config/config.yaml`).

The original purpose of `test/analysis/input` was to archive the output structure of SPRAS. This structure has not fundamentally changed, but we can no longer store that structure inside the repository easily, as with this PR, the hashes change when the SPRAS commit revision changes (as explained in the original issue description). Since we did not test against this entire structure, but only against cytoscape and summary files, we could safely remove this file.

Now, these tests do Snakemake runs instead (also as a side-goal to encourage integration testing #165).

File: spras/config/config.py [hunk 1]
```diff
@@ -27,6 +30,24 @@
...
+@functools.cache
+def spras_revision() -> str:
+    """
+    Gets the revision of the current SPRAS repository.
+    If this file is inside a `.git` repository, this uses the revision hash.
+    Otherwise, this uses the package version.
+    """
+    # Check if we're inside a git repository
+    try:
+        return subprocess.check_output(["git", "rev-parse", "--short", "HEAD"], encoding='utf-8').strip()
```
Comment (agitter) [inline spras/config/config.py]: Won't this return successfully if we are in any git repository, not just the spras repo? I could have spras and it's dependencies installed within a different project's repo.

File: spras/config/config.py [hunk 2]
```diff
@@ -27,6 +30,24 @@
...
+    except subprocess.CalledProcessError:
+        # Use the current package version instead
+        # https://stackoverflow.com/a/75100875/7589775
+        return f"v{importlib.metadata.version('spras').replace('.', '_')}"
+
+def attach_spras_revision(label: str) -> str:
+    return f"{label}_{spras_revision()}"
```
Comment (agitter) [inline spras/config/config.py]: I'm thinking through whether there are other ways to get this same behavior without making filenames longer. The subdirectory names that already follow the <dataset>-<algorithm>-params-<hash> pattern are long already, and now we're extending them. The only other idea is to use subdirectories instead, which isn't necessarily an improvement.

File: test/analysis/input/egfr.yaml [hunk 3]
```diff
@@ -1,92 +1,28 @@
-# The length of the hash used to identify a parameter combination
```
Comment (agitter) [inline test/analysis/input/egfr.yaml]: Why did the parameter choices change here? That is unrelated to adding the revision, isn't it?

**Review by agitter — COMMENTED** [review]
I finished another partial revision. I still haven't thought about the testing implications carefully.

File: spras/config/config.py (hunk 2 above)
Comment (tristan-f-r) [reply spras/config/config.py]: This is a little concerning, though thanks to #434, I'm not too worried about files being the primary interface for organizing SPRAS output. We should still document this file directory naming once we have actual SPRAS workflow documentation.

File: spras/config/config.py (hunk 1 above)
Comment (tristan-f-r) [reply spras/config/config.py]: Actually, I took a slightly different approach, since:
- I have now learned that there is no git repository attached to pip-vcs-based installs.
- I want to make sure that people don't install SPRAS wrong, so this function is now user friendly at the cost of more code.

We do check that the `git` call is the correct one via `pyproject.toml`, but if that fails, we _instead_ rely on the hash of the associated RECORD file (which is just a checksum file of every single other dependency in the project).

File: spras/config/config.py [hunk 4]
```diff
@@ -27,6 +33,59 @@
...
+                raise RuntimeError(f"The git top-level `{pyproject_path}` is not the SPRAS pyproject.toml. {clone_tip}")
+
+            return subprocess.check_output(
+                ["git", "rev-parse", "--short", "HEAD"],
+                encoding='utf-8',
+                cwd=project_directory
+            ).strip()
+        except FileNotFoundError as err:
+            # pyproject.toml wasn't found during the `read_text` call
+            raise RuntimeError(f"The git top-level {pyproject_path} wasn't found. {clone_tip}") from err
```
Comment (agitter) [inline spras/config/config.py]: It looks like a user would hit this RuntimeError if they pip install spras and then try to use it later in a different git repo with a different top-level name. That should be a valid use case. 


What happens if someone is in a spras repo and has the package pip installed with conflicting versions?

The various control flow paths here are pretty complicated, and I'm not confident we've covered every corner case.

**Review by agitter — COMMENTED** [review]
A few more comments. I still haven't looked through all the test code.

**Comment by agitter** [comment]
As a follow up to our meeting discussion, I'm wondering if this type of output file versioning should be optional. Then when running in CHTC and writing to OSDF (or running locally and opting in) it could be enabled. By making it opt in, we would have simpler filenames by default and ensure the user knows they have to install and run SPRAS a specific way for this feature to work.

**Comment by tristan-f-r** [comment]
That makes the most sense to me as well :+1:

File: config/config.yaml [hunk 5]
```diff
@@ -3,6 +3,15 @@
...
 hash_length: 7
 
+# If enabled, this tags all files with their local file version.
+# Most files do not have a specific version, and by default, this will be the hash of
+# all the SPRAS files in the PyPA installation. This option will not work if SPRAS was not installed
+# in a PyPA-compliant manner (PyPA-compliant installations include but are not limited to pip, poetry, uv, conda, pixi.)
+#
+# By default, this is disabled, as it can make output file names confusing. Here, it's set to true since we use this
+# configuration file for testing.
+osdf_immutable: true
```
Comment (tristan-f-r) [inline config/config.yaml]: This is a little annoying. We use this config for testing, so it's nice to enable this, but this is also our documentation config. I can write some extra code to enable this during testing, but that seems strange as well.

**Review by agitter — COMMENTED** [review]
This is getting closer and the scope looks better now that the labeled outputs are optional. I still will need a final end-to-end pass because I've been reviewing in chunks.

File: spras/config/config.py [hunk 6]
```diff
@@ -28,62 +26,47 @@
...
-    2. If SPRAS was installed via a PyPA-compliant package manager, we use the hash of the RECORD file (https://packaging.python.org/en/latest/specifications/recording-installed-packages/#the-record-file).
-        which contains the hashes of all installed files to the package.
+    Gets the current revision of SPRAS.
+
+    A few notes:
+    - This is not dependent on the SPRAS version nor the git commit, but rather solely on the PyPA RECORD file,
+    (https://packaging.python.org/en/latest/specifications/recording-installed-packages/#the-record-file), which contains
+    hashes of all files associated with the package distribution [other than itself], and is also included in the package distribution.
+    - This means that, when developing SPRAS, spras_revision will be updated when spras is initially installed, but not during the middle
+    of development.
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply spras/config/config.py]: Indeed. I'll clarify that 👍

File: spras/config/config.py [hunk 7]
```diff
@@ -28,62 +26,47 @@
...
+def attach_spras_revision(osdf_immutable: bool, label: str) -> str:
+    """
+    Attaches the SPRAS revision to a label.
+    This function signature may become more complex as specific labels get versioned.
+
+    @param label: The label to attach the SPRAS revision to.
+    @param osdf_immutable: if False, this function is equivalent to `id`.
+    """
+    if osdf_immutable is False: return label
     return f"{label}_{spras_revision()}"
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply spras/config/config.py]: Yes! It did take me some time to figure out why I did that, though - this is very needed documentation.

File: spras/config/config.py [hunk 8]
```diff
@@ -28,62 +26,47 @@
...
-    Gets the revision of the current SPRAS repository. This function is meant to be user-friendly to warn for bad SPRAS installs.
-    1. If this file is inside the correct `.git` repository, we use the revision hash. This is for development in SPRAS as well as SPRAS installs via a cloned git repository.
-    2. If SPRAS was installed via a PyPA-compliant package manager, we use the hash of the RECORD file (https://packaging.python.org/en/latest/specifications/recording-installed-packages/#the-record-file).
-        which contains the hashes of all installed files to the package.
+    Gets the current revision of SPRAS.
+
+    A few notes:
+    - This is not dependent on the SPRAS version nor the git commit, but rather solely on the PyPA RECORD file,
+    (https://packaging.python.org/en/latest/specifications/recording-installed-packages/#the-record-file), which contains
+    hashes of all files associated with the package distribution [other than itself], and is also included in the package distribution.
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply spras/config/config.py]: I figured it out, and I documented this behavior at https://github.com/Reed-CompBio/spras/pull/320/commits/9ab902aee937aacaf6a9c2c2aa0d6a5f9546edb8. This actually ends up being expected :+1:

File: Snakefile [hunk 9]
```diff
@@ -209,14 +209,15 @@ checkpoint prepare_input:
         # Use the algorithm's generate_inputs function to load the merged dataset, extract the relevant columns,
         # and write the output files specified by required_inputs
         # The filename_map provides the output file path for each required input file type
-        filename_map = {input_type: SEP.join([out_dir, 'prepared', f'{wildcards.dataset}-{wildcards.algorithm}-inputs', f'{input_type}.txt']) for input_type in runner.get_required_inputs(wildcards.algorithm)}
-        runner.prepare_inputs(wildcards.algorithm, input.dataset_file, filename_map)
+        filename_map = {input_type: SEP.join([out_dir, 'prepared', f'{wildcards.dataset}-{wildcards.algorithm}-inputs', f'{input_type}.txt']) for input_type in runner.get_required_inputs(detach_spras_revision(_config.config.immutable_files, wildcards.algorithm))}
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply Snakefile]: I'm wary of making algorithm hashes special for parameter logs, especially when we will want to introduce a more permanent version of an algorithm hash as per #464.

**Review by ntalluri — COMMENTED** [review]
I ran the code locally, the hashes aren't as long as I thought, still very manageable.  I also left so changes and questions.

File: spras/config/revision.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+import functools
+import hashlib
+import importlib.metadata
+import sysconfig
+from pathlib import Path
+
+
+@functools.cache
+def spras_revision() -> str:
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply spras/config/revision.py]: `@functools.cache` guarantees that it is called only once.

File: config/config.yaml [hunk 11]
```diff
@@ -3,6 +3,15 @@
 # The length of the hash used to identify a parameter combination
 hash_length: 7
 
+# If enabled, this tags all output files with a SPRAS 'revision version'.
+# By default, this will be the hash of all the SPRAS files in the PyPA installation. This option will not work if SPRAS was not installed
+# in a PyPA-compliant manner (PyPA-compliant installations include but are not limited to pip, poetry, uv, conda, pixi.)
+# For some files, the 'SPRAS revision' may be tied to the specific format version that file is on.
+#
+# By default, this is disabled, as it can make output file names confusing. Here, it's set to true since we use this
+# configuration file for testing.
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply config/config.yaml]: That's important clarification.

File: config/config.yaml [hunk 12]
```diff
@@ -3,6 +3,15 @@
 # The length of the hash used to identify a parameter combination
 hash_length: 7
 
+# If enabled, this tags all output files with a SPRAS 'revision version'.
+# By default, this will be the hash of all the SPRAS files in the PyPA installation. This option will not work if SPRAS was not installed
+# in a PyPA-compliant manner (PyPA-compliant installations include but are not limited to pip, poetry, uv, conda, pixi.)
+# For some files, the 'SPRAS revision' may be tied to the specific format version that file is on.
+#
+# By default, this is disabled, as it can make output file names confusing. Here, it's set to true since we use this
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply config/config.yaml]: A good chunk of this comment is to explain that this file is used half as a testing artifact and half as documentation.
